messaging/src/main/java/org/axonframework/messaging/MessageStreamUtils.java

```diff
@@ -0,0 +1,212 @@
+     */
+    public static <M extends Message<?>, R> CompletableFuture<R> reduce(MessageStream<M> source,
+                                                                        R identity,
+                                                                        BiFunction<R, MessageStream.Entry<M>, R> accumulator) {
+        Reducer<M, R> reducer = new Reducer<>(source, identity, accumulator);
+        source.onAvailable(reducer::process);
+        return reducer.completableFuture();
+    }
+
+    /**
+     * Returns a CompletableFuture that completes with the first element from the given {@code source}.
+     * <p>
+     * If the given source completed without producing any elements, the returned CompletableFuture will either complete
+     * with a {@code null} result, of the source completed normally, or exceptionally if the source completed with an
+     * error.
+     * <p>
+     * Once the first element is read from the source, it is automatically closed and any subsequent elements in the
+     * {@code source} are ignored.
+     *
+     * @param source The source to read the first message from
+     * @param <M>    The type of Message produced by the stream
+     * @return a CompletableFuture that completes with the first message from the stream
+     */
+    public static <M extends Message<?>> CompletableFuture<MessageStream.Entry<M>> firstAsCompletableFuture(
+            MessageStream<M> source) {
+        FirstResult<M> firstResult = new FirstResult<>(source);
+        source.onAvailable(firstResult::process);
+        return firstResult.completableFuture();
+    }
+
+    private static class FluxTask<M extends Message<?>> {
+
+        private final AtomicBoolean processingGate = new AtomicBoolean(false);
+        private final MessageStream<M> source;
+        private final FluxSink<MessageStream.Entry<M>> emitter;
+
+        public FluxTask(MessageStream<M> source, FluxSink<MessageStream.Entry<M>> emitter) {
+            this.source = source;
+            this.emitter = emitter;
+        }
+
+        public void process() {
+            boolean continueOnCurrentThread = true;
+            while (continueOnCurrentThread && !processingGate.getAndSet(true)) {
+                try {
+                    while (emitter.requestedFromDownstream() > 0 && source.hasNextAvailable()) {
+                        source.next().ifPresent(emitter::next);
+                    }
+                    if (source.isCompleted()) {
+                        source.error().ifPresentOrElse(emitter::error, emitter::complete);
+                    }
+                } catch (Exception e) {
+                    emitter.error(e);
+                    source.close();
+                } finally {
+                    processingGate.set(false);
+                }
+                continueOnCurrentThread = emitter.requestedFromDownstream() > 0 && source.hasNextAvailable();
```

Comment:
No, this check must be performed after the gate, and is to ensure that no additional tasks are waiting to be processed when the gate is released. It may be that another thread that has been signalled, was refused at the gate.

Reply:
Cool, thanks for clarifying!